### ISSUE TITLE
Conv1D Multichannel broken

### ISSUE THREAD

**Issue by jamarshon** [pytorch/pytorch#23042]
## Steps to produce
I run the following code with two setups. Basically, it is convolution on a multichannel tensor (2, 8024) versus running convolution on the same tensor but with only one channel (1, 8024).
 
```
def conv(i, w):
	conv_stride = 2
	num_channels = i.size(0)
	return torch.nn.functional.conv1d(
		i.unsqueeze(0), w.repeat(num_channels, 1, 1),
        stride=conv_stride, groups=num_channels).squeeze(0)

torch.manual_seed(0)
i1 = torch.rand((2, 8024))*1e2
w1 = torch.rand(25)*1e2

conv1 = conv(i1, w1)  # (2, 4000)
conv2 = conv(i1[0,:].unsqueeze(0), w1)  # (1, 4000)

print('conv error %.16f' % ((conv1[0,:] - conv2[0,:]).abs().max()))
```
On the first setup, I use
```
conda create --name six python=3.7
conda activate six
conda install pytorch-nightly=1.2.0.dev20190615 -c pytorch
```
and get `conv error 0.0000000000000000` as expected

For the second setup, I use
```
conda create --name seven python=3.7
conda activate seven
conda install pytorch-nightly=1.2.0.dev20190718 -c pytorch
```

and get `conv error 0.0234375000000000` which is not correct.

## Environment
Collecting environment information...
PyTorch version: 1.2.0.dev20190718
Is debug build: No
CUDA used to build PyTorch: Could not collect
OS: Mac OSX 10.14.5
GCC version: Could not collect
CMake version: version 3.14.0
Python version: 3.7
Is CUDA available: No
CUDA runtime version: Could not collect
GPU models and configuration: Could not collect
Nvidia driver version: 1.1.0
cuDNN version: Could not collect
Versions of relevant libraries:
[pip] numpy==1.16.4
[pip] torch==1.2.0.dev20190718
[conda] mkl                       2019.4                      233  
[conda] pytorch-nightly           1.2.0.dev20190718         py3.7_0    pytorch

**Comment by cpuhrsch** [comment]
@jamarshon - can you reduce the input size (smaller than (1, 8024)) and use deterministic numbers (say arange or ones) to reproduce this? Is there a certain size or certain data that causes this discrepancy? Is it unique to GPU or CPU?

**Comment by jamarshon** [comment]
A smaller example of size 20 with `conv error 0.0004882812500000`
```
def conv(i, w):
	conv_stride = 2
	num_channels = i.size(0)
	return torch.nn.functional.conv1d(
		i.unsqueeze(0), w.repeat(num_channels, 1, 1),
        stride=conv_stride, groups=num_channels).squeeze(0)

input1 = torch.tensor([
		[49.6257, 76.8222,  8.8477, 13.2030, 30.7423, 63.4079, 49.0093, 89.6445,
         45.5628, 63.2306, 34.8893, 40.1717,  2.2326, 16.8859, 29.3888, 51.8522,
         69.7668, 80.0011, 16.1029, 28.2269],
        [68.1609, 91.5194, 39.7100, 87.4156, 41.9408, 55.2907, 95.2738,  3.6165,
         18.5231, 37.3417, 30.5100, 93.2000, 17.5910, 26.9834, 15.0680,  3.1720,
         20.8130, 92.9799, 72.3109, 74.2336]])
input2 = input1[0, :].unsqueeze(0)
weight = torch.tensor([52.6296, 24.3658])
conv1 = conv(input1, weight)  # (2, 10)
conv2 = conv(input2, weight)  # (1, 10)
print('conv1', conv1)
print('conv2', conv2)
print('conv error %.16f' % ((conv1[0,:] - conv2[0,:]).abs().max()))
```

**Comment by cpuhrsch** [comment]
This doesn't reproduce on an older checkout (2019-04-18), but does reproduce on a newer one - @jamarshon could you bisect this error and assign the creator of the offending commit?

**Comment by rrkarim** [comment]
@cpuhrsch Maybe it is because of #19204? Take a look at the `bool use_mkldnn` def.

**Comment by cpuhrsch** [comment]
Here is a more succinct example based on Jason's script

```
$ cat /tmp/error.py
import torch
def conv(i, w):
    conv_stride = 2
    num_channels = i.size(0)
    return torch.nn.functional.conv1d(
        i.unsqueeze(0), w.repeat(num_channels, 1, 1),
        stride=conv_stride, groups=num_channels).squeeze(0)

def try_run(size):
    num = 1
    for s in size:
        num *= s
    input1 = torch.arange(num)
    input1 = input1.view(size).float()
    input1 = input1 / 6
    print(input1)
    input2 = input1[0, :].unsqueeze(0)
    weight = torch.tensor([1., 2.]).mul(5)
    conv1 = conv(input1, weight)  # (2, 10)
    conv2 = conv(input2, weight)  # (1, 10)
    print(conv1[0, :])
    print(conv2[0, :])
    return (conv1[0,:] - conv2[0,:]).abs().max()
print(try_run((2, 6)))

$ python /tmp/error.py
tensor([[0.0000, 0.1667, 0.3333, 0.5000, 0.6667, 0.8333],
        [1.0000, 1.1667, 1.3333, 1.5000, 1.6667, 1.8333]])
tensor([ 1.6667,  6.6667, 11.6667])
tensor([ 1.6667,  6.6667, 11.6667])
tensor(9.5367e-07)
```

**Comment by jamarshon** [comment]
@rrkarim since I said it worked in setup case #1 (pytorch-nightly=1.2.0.dev20190615), I don't think the Apr 27 merge should have affected it

**Comment by jamarshon** [comment]
The first broken setup is `pytorch-nightly=1.2.0.dev20190618`. `pytorch-nightly=1.2.0.dev20190617` works as expected (`conv error 0.0000000000000000`)

**Comment by rrkarim** [comment]
@jamarshon yeah I see it.

**Comment by soumith** [comment]
hmmm im2col / vol2col PR might be a potential one: https://github.com/pytorch/pytorch/pull/21769
updating MKL-DNN bridge might be another (suspect this one): https://github.com/pytorch/pytorch/pull/21854

**Comment by jamarshon** [comment]
```
>>> torch.__version__
'1.2.0a0+5237835'
conv error 0.0000000000000000

>>> torch.__version__
'1.2.0a0+7f057f0'
conv error 0.0004882812500000
```
looks like it's commit 7f057f0 which is https://github.com/pytorch/pytorch/pull/21854

**Comment by soumith** [comment]
cc: @gujinghui @Jianhui-Li @uyongw @jgong5 @XiaobingSuper

Please look into this asap. It is **yet** another convolution correctness bug.

**Comment by Jianhui-Li** [comment]
@soumith We are looking into this.

**Comment by soumith** [comment]
as a post-mortem after this bug has been resolved, can we significantly write more test cases for MKL-DNN, including randomized testing. Time and again we get flagged correctness bugs, and I think comprehensive tests will get to the bottom of this.

**Comment by Jianhui-Li** [comment]
@soumith Thanks for the feedback. Yes I do agree with you that more tests are needed. Let me follow up on this.

**Comment by emfomenk** [comment]
Hi,

Intel MKL-DNN developer here.
TL;DR; I was able to reproduce the results of both original and simplified examples. And I am sure the difference is caused not by a bug in the library but by the finite precision of the float point arithmetic (in other words rounding issues).

Deep into details...

The simplest way to confirm that is to use integer inputs that fit into mantissa of `float`, which is 23-bit wide. That works perfectly fine for operations like convolution that doesn't divide anything or perform any fancy functions like `tanh`. I applied a simple *fix* [2] to both reproducers and they both pass [1]. Using integers in validation allows avoiding false-negatives like in these cases.

If approach with integers is not applicable for whatever reason, that you should expect that the computed result might diverge from the *true* one (I also want to note that the reference implementation is not a *true* result, since it was also computed in finite precision). For simple operations like `a+b` or `a*b` one might expect that the relative difference is no more that 1e-7 (assuming `f32` data type). Note, that in the original `/tmp/error.py` that was exactly the case: `9.5367e-07 / 11.6667 < 1e-7`. So even from that perspective we could conclude that the result is most like correct.

For complex operations like GEMM, Conv, BatchNorm and any others that require a chain of computations for each output result, the relative error might grow. It is really complicated task to come up with robust and reliable way to test things. Intel MKL-DNN uses quite advanced algorithms to validate these primitives. For convolutions we try using integers whenever possible. For BatchNorm we prepare special data that allows avoiding rounding issues during computations. For GEMM we could use integers as well, but so far we use some criteria like `rel_err < 1e-7 * log(K_dim)`, which is a reasonable error bound for GEMM.

---

[1] Before and after the *fix*.

``` sh
$ python3 orig.py && echo && python3 error.py
conv error 0.0234375000000000

tensor([[0.0000, 0.1667, 0.3333, 0.5000, 0.6667, 0.8333],
        [1.0000, 1.1667, 1.3333, 1.5000, 1.6667, 1.8333]])
tensor([ 1.6667,  6.6667, 11.6667])
tensor([ 1.6667,  6.6667, 11.6667])
tensor(9.5367e-07)

$ git apply fix.patch
$ python3 orig.py && echo && python3 error.py
conv error 0.0000000000000000

tensor([[ 0.,  1.,  2.,  3.,  4.,  5.],
        [ 6.,  7.,  8.,  9., 10., 11.]])
tensor([10., 40., 70.])
tensor([10., 40., 70.])
tensor(0.)
```

[2] the `fix.patch`
``` patch
commit 3f3094bbd34fd72d377af0a062c4b6e7719d0343 (HEAD -> master)
Author: emfomenk <evarist.m.fomenko@intel.com>
Date:   Fri Jul 19 17:19:38 2019 -0700

    make tests rounding-error resistant

diff --git a/error.py b/error.py
index 93cd2db..2201726 100644
--- a/error.py
+++ b/error.py
@@ -12,7 +12,7 @@ def try_run(size):
         num *= s
     input1 = torch.arange(num)
     input1 = input1.view(size).float()
-    input1 = input1 / 6
+    input1 = input1 # / 6
     print(input1)
     input2 = input1[0, :].unsqueeze(0)
     weight = torch.tensor([1., 2.]).mul(5)
diff --git a/orig.py b/orig.py
index 1e8e14e..fcf0cfd 100644
--- a/orig.py
+++ b/orig.py
@@ -8,8 +8,11 @@ def conv(i, w):
         stride=conv_stride, groups=num_channels).squeeze(0)

 torch.manual_seed(0)
-i1 = torch.rand((2, 8024))*1e2
-w1 = torch.rand(25)*1e2
+i1 = torch.rand((2, 8024))*1e3
+w1 = torch.rand(25)*1e3
+
+i1 = torch.round(i1)
+w1 = torch.round(w1)

 conv1 = conv(i1, w1)  # (2, 4000)
 conv2 = conv(i1[0,:].unsqueeze(0), w1)  # (1, 4000)
```

**Comment by emfomenk** [comment]
P.S. If replace `/6` by `/8` in the `error.py` the problem will go away as well, because division by 8 is *exact* and keep a lot of zeros in trailing bits in mantissa, while division by `/6` cannot be always done accurately with radix=2, so you end up having tensor whose values have a lot of non-zero bits in mantissa. And that's exactly why the rounding issues appear in this case.

**Comment by jamarshon** [comment]
@emfomenk Thanks for the reply! So even with input as integers and within 23 bits, I get 
`conv error 256.0000000000000000` which seems high. Is this expected? It just seems strange that commit 5237835a1 does not face any precision issues whereas the commit after it does (7f057f00c)
```
torch.manual_seed(0)
i1 = torch.rand((2, 20))*1e5
w1 = torch.rand(2)*1e5

i1 = i1.round()
w1 = w1.round()
conv1 = conv(i1, w1)  # (2, 10)
conv2 = conv(i1[0,:].unsqueeze(0), w1)  # (1, 10)

print('conv error %.16f' % ((conv1[0,:] - conv2[0,:]).abs().max()))
```

**Comment by emfomenk** [comment]
Hi @jamarshon,

I guess the output doesn't fit into 23 bits (i.e. greater than 8 * 10**7). If that's the case, then there is no difference in between integers and original floating point numbers -- for both all mantissa bits will be used, and rounding issues might occur.

**Comment by emfomenk** [comment]
oK, I actually just realized that since both the input and weights are of magnitude `1e5` then the output of would be of magnitude `1e10` which way greater than `8e6` :)
So yes, this behavior is expected.

BTW, note that the relative error is again less than `1e-7`:

```
$ python3 orig.py
conv error 256.0000000000000000
conv rel error 4.554235e-08  <--- relative error less than 1e-7
```

``` patch
@@ -8,8 +8,8 @@ def conv(i, w):
         stride=conv_stride, groups=num_channels).squeeze(0)

 torch.manual_seed(0)
-i1 = torch.rand((2, 8024))*1e3
-w1 = torch.rand(25)*1e3
+i1 = torch.rand((2, 20))*1e5
+w1 = torch.rand(2)*1e5

 i1 = torch.round(i1)
 w1 = torch.round(w1)
@@ -18,3 +18,4 @@ conv1 = conv(i1, w1)  # (2, 4000)
 conv2 = conv(i1[0,:].unsqueeze(0), w1)  # (1, 4000)

 print('conv error %.16f' % ((conv1[0,:] - conv2[0,:]).abs().max()))
+print('conv rel error %e' % (((conv1[0,:] - conv2[0,:]).abs().max()) / (conv1[0,:]).abs().max()))
```

**Comment by uyongw** [comment]
@jamarshon Thanks for reporting the issue. As @emfomenk already mentioned, we think it is a numerical accuracy issue, or an issue of FP arithmetic numerical reproducibility, other than a correctness issue.
> It just seems strange that commit 5237835 does not face any precision issues whereas the commit after it does (7f057f0)

The commit 7f057f0 causes some group convolution to use a different kernel implementation than others. While from the perspective of numerical accuracy, results of different math kernel implementation are not fully consistent. The variance can be caused by 

- Different accumulation order caused by matrix partition and thread parallelism, as we know FP addition is not associative due to rounding error.
- Different FP instruction sequence used. For example, fused-multiply-add (FMA) instruction will have better accuracy than multiply followed by add, because it computes a*b+c in only one rounding step.
- Different rounding mode.

Thus when you compare the results of different kernel path there is a small delta observed. Following up, whether to force numerical reproducibility from framework or acceleration library is still a topic we need further investigation. Any suggestions or comments are welcome.

**Comment by XiaobingSuper** [comment]
@soumith, perhaps this is numerical accuracy issue, because there has also big variance between GPU path and cpu native path(not using mkldnn by disable mkldnn in [L150](https://github.com/pytorch/pytorch/blob/b8c8977be797cb16354aa97c3fcdc910d3e9e5c5/aten/src/ATen/native/Convolution.cpp#L150)) , but has a little variance between gpu and mkldnn.  The test script is:
```
import torch

def conv(i, w):
    conv_stride = 2
    num_channels = i.size(0)
    return torch.nn.functional.conv1d(
            i.unsqueeze(0), w.repeat(num_channels, 1, 1),
            stride=conv_stride, groups=num_channels).squeeze(0)

torch.manual_seed(0)
i1 = torch.rand((2, 8024))*1e2
w1 = torch.rand(25)*1e2
i2 = i1.cuda()
w2 = w1.cuda()

# CPU path 
conv1 = conv(i1, w1)  # (2, 4000)
conv2 = conv(i1[0,:].unsqueeze(0), w1)  # (1, 4000)
print('cpu conv error %.16f' % ((conv1[0,:] - conv2[0,:]).abs().max()))

# GPU path
conv11 = conv(i2, w2)  # (2, 4000)
conv22 = conv(i2[0,:].unsqueeze(0), w2)  # (1, 4000)
print('gpu conv error %.16f' % ((conv11[0,:] - conv22[0,:]).abs().max()))

#compare cpu and gpu path
print("compare cpu and gpu result")
print('first conv error %.16f' % ((conv11 - conv1.cuda()).abs().max()))
print('second conv error %.16f' % ((conv22 - conv2.cuda()).abs().max()))
``` 
If disable mkldnn, the ouput log is :
```
cpu conv error 0.0000000000000000
gpu conv error 0.0000000000000000
compare cpu and gpu result
first conv error 0.0234375000000000
second conv error 0.0234375000000000
```
else using mkldnn, the output log is:
```
cpu conv error 0.0000000000000000
gpu conv error 0.0000000000000000
compare cpu and gpu result
first conv error 0.0000000000000000
second conv error 0.0000000000000000
```
The test CPU device is **skx-8180** and GPU device is **P100**.  thanks！

**Comment by soumith** [comment]
@emfomenk thanks so much for getting to the bottom of it so quickly. everything you said makes sense.

**Comment by Jianhui-Li** [comment]
@soumith We had a post-mortem for the recent issues on mkl-dnn conv.  MKL-DNN is committed to continuously improve validation quality and test coverage. This is an ongoing process.  We are planning systematic code coverage analysis to monitor quality of the efforts.  

Thanks for understanding this specific issue is not a defect as FP arithmetic does not guarantee bitwise identical results.
